packages/node/src/utils/kyve/kyve.ts

```diff
@@ -0,0 +1,529 @@
+  }
+
+  private async getBundleById(bundleId: number): Promise<BundleDetails> {
+    return (this.cachedBundleDetails[bundleId] ??= (() => {
+      logger.debug(`getBundleId ${bundleId}`);
+      return this.lcdClient.kyve.query.v1beta1.finalizedBundle({
+        pool_id: this.poolId,
+        id: bundleId.toString(),
+      }) as Promise<BundleDetails>;
+    })());
+  }
+
+  private async getResolvedBundleDetails(): Promise<BundleDetails[]> {
+    return Promise.all(Object.values(this.cachedBundleDetails));
+  }
+
+  private async getLatestBundleId(): Promise<number> {
+    return (
+      parseDecimal(
+        (
+          await this.lcdClient.kyve.query.v1beta1.finalizedBundles({
+            pool_id: this.poolId,
+            index: '1',
+            pagination: {
+              limit: '1',
+            },
+          })
+        ).pagination.total,
+      ) - 1
+    ); // bundle id starts from 0
+  }
+
+  private async getBundleId(height: number): Promise<number> {
+    const lowestCacheHeight = Object.keys(this.cachedBundleDetails);
+
+    let low =
+      lowestCacheHeight.length > 0
+        ? Math.min(...lowestCacheHeight.map(parseDecimal))
+        : -1;
+    let high = await this.getLatestBundleId();
+
+    while (low <= high) {
+      const mid = Math.floor((low + high) / 2);
+      const midBundle = await this.getBundleById(mid);
+
+      const fromKey = parseDecimal(midBundle.from_key);
+      const toKey = parseDecimal(midBundle.to_key);
+
+      if (height >= fromKey && height <= toKey) {
+        return mid;
+      }
+
+      if (height > toKey) {
+        low = mid + 1;
+      } else {
+        high = mid - 1;
+      }
+    }
+    throw new Error(`No suitable bundle found for height ${height}`);
+  }
+
+  private findBlockByHeight(
+    height: number,
+    fileCacheData: KyveBundleData[],
+  ): KyveBundleData | undefined {
+    return fileCacheData.find(
+      (bk: KyveBundleData) => bk.key === height.toString(),
+    );
+  }
+
+  private async getBundleFromCache(
+    height: number,
+  ): Promise<BundleDetails> | undefined {
+    const bundles = await this.getResolvedBundleDetails();
+    return bundles.find(
+      (b) =>
+        parseDecimal(b.from_key) <= height && height <= parseDecimal(b.to_key),
+    );
+  }
+
+  private async updateCurrentBundleAndDetails(
+    height: number,
+  ): Promise<KyveBundleData[]> {
+    let bundle = await this.getBundleFromCache(height);
+
+    if (!bundle) {
+      const bundleIds = Object.keys(this.cachedBundleDetails);
+
+      let bundleId: number;
+      if (bundleIds.length === 0) {
+        bundleId = await this.getBundleId(height);
+      } else {
+        //
+        const cachedBundles = await this.getResolvedBundleDetails();
+        const nearestBundle = cachedBundles.filter(
+          (b) => parseDecimal(b.to_key) < height,
+        );
+
+        bundleId =
+          Math.max(...nearestBundle.map((b) => parseDecimal(b.id))) + 1;
```

Comment:
There's no guarantee that the next bundle is last bundle + 1. What happens if there is a large jump in block number?